Add Read Signal to Noise Ratio in the read stats.

ppapasaikas opened the pull request:
Briefly total read variance decomposes to Signal and Noise variance. 
Noise variance is estimated as 0.5*var(Dx) where Dx are lagged deltas, and a floor noise variance is applied.
Then SignalVar=TotalVar - NoiseVar and SNR=log2(SignalVar/NoiseVar).
This metric typically shows a very long tail of low SNR values corresponding to spurious  "flat", typically lowly Methylated reads and works better then existing metrics for identifying and removing those reads (about 10% in our libraries).


Make also corresponding  changes in plotReadStats and filterReads + some minor cleanup changes in calcReadStats

ppapasaikas committed:
Add Read Signal to Noise Ratio in the read stats. Make corresponding changes in plotReadStats and filterReads + some minor cleanup changes in calcReadStats

ppapasaikas committed:
Add tests for new metrics

csoneson committed:
Rerender documentation

csoneson committed:
Import stats::dist

csoneson committed:
Import stats::coef

csoneson committed:
Document minSNR in plotReadStats

csoneson committed:
Reduce number of cores used in countStatePairs example

csoneson committed:
Minor style edits for consistency

csoneson committed:
Limit number of cores in tests

csoneson committed:
Adapt unit tests

csoneson commented:
There's still one test failing - it's not clear to me why this result is changing because of the changes introduced here 🤔 

Also just wondering if there would be a way to simplify/avoid the caching part by pre-calculating the results of `.estimate_snr_probList` if any of the three new statistics should be calculated, and then sending that output to the respective functions?

ppapasaikas commented:
Thank you. I'll try to figure out what triggers the fail and also try to avoid caching.

ppapasaikas committed:
Remove caching requirement for SNR stats

ppapasaikas committed:
simplify do.call code in calcReadStats

ppapasaikas committed:
Add back discrete SNR, SignalVar, NoiseVar functions to facilitate testing

ppapasaikas committed:
Removed now redundant Lag1DModProb metric.
For the lagged diffs distance of observations is based on the pos(rowRanges(se)) of the NNAindices rather than on NNAindices themselves.

ppapasaikas committed:
Missing import and correct number of expected stats for calcReadStats

ppapasaikas committed:
Add functions for SNR estimation in the scanGenome framework

ppapasaikas committed:
update documentation

ppapasaikas committed:
Account for multiple samples in estimateNoiseParsWindows.
Style changes.
Thresholds for noise estimation

ppapasaikas committed:
Update documentation

ppapasaikas committed:
Thresholding changes

mbstadler committed:
Merge remote-tracking branch 'origin/devel' into feature/ReadSNR

mbstadler committed:
use cli::cli_warn instead of warning

mbstadler committed:
catch warnings in unit tests

mbstadler committed:
make SNRstats a global variable (consistent with other defaultReadStats and allReadStats)

mbstadler committed:
consolidate redundant imports

mbstadler committed:
remove obsolete BiocGenerics::

mbstadler committed:
add missing import of MulticoreParam

mbstadler committed:
add missing imports

mbstadler committed:
use .assert* functions for argument checking

mbstadler committed:
rename variable to avoid clash with stats::coef

mbstadler committed:
use \code instead of backticks in roxygen2 documentation

mbstadler committed:
minor formatting changes

mbstadler committed:
update docs

mbstadler commented:
- a brief description of the background model would be great, at least for us internally?
- there are many hard-coded/hidden cutoffs in the code. Should maybe some of them be exposed as arguments, so that we are aware of them and can adjust if necessary (for example the constants in `min_diffs <- max(16L, floor(0.05 * length(x)))` or `d_cut <- max(10, quantile(depth, 0.05, na.rm = TRUE))` in `.estimate_snr_vec`, or `d_cut <- max(10, quantile(depthVals, 0.05, na.rm = TRUE))` and the six lower and upper quantiles used for pre-fit filtering in `estimateNoiseParsWindows`)?
-  alternatively, could a robust `lm` be an alternative to the standard `lm` that would eliminate many or all of these hard-coded parameters?
- should we rename `nWindows` to `nWindowsPerChromosome` in `estimateNoiseParsWindows()`, to better reflect its meaning?
- would it be an option to lift-out the plotting functionality in `estimateNoiseParsWindows()` to a separate function, consistent with other parameter estimation/plotting functions (e.g. `calcReadStats`, `plotReadStats`)?
- should we make the plot in `estimateNoiseParsWindows()` using `ggplot2` instead of base graphics, to be consistent with all other plotting?
- we should add some unit tests (we currently have ~200 missed lines in scanGenome.R)
- we should also add `minSNR` parameter and filtering to `filterReadsBam`, which is in every other way consistent with `filterReads`
- we should bump the version and update `NEWS.md`

csoneson committed:
Fix some spaces

csoneson committed:
Add back some formatting edits for consistency

mbstadler committed:
merge devel (fix filtering of reads without modified-base calls)

Merge branch 'devel' into feature/ReadSNR

# Conflicts:
#	R/filterReads.R

ppapasaikas committed:
Documentation corrections.
Rework setup of filtering parameters so they are transparent to user but also validated

ppapasaikas committed:
Decouple plotting from background model estimation function.

ppapasaikas committed:
Separate plotting functionality in estimateNoiseParsWindows
Update documentation

ppapasaikas committed:
Addition of SNR filtering option in filterReadsBam

ppapasaikas committed:
Updated cpp files for SNR option

ppapasaikas committed:
Switch obeject returned by NoiseFilterParam to a simple list

csoneson committed:
Merge branch 'devel' into feature/ReadSNR

ppapasaikas committed:
Return fitted coefficients of bakcground noise model.

ppapasaikas committed:
Add the option to pass background noise model coefficients from a previous calculation.
Remove unnecessary overloading in cpp function definition.

ppapasaikas committed:
k should allow 0 (0 means exactly adjacent)

ppapasaikas committed:
Add tests for SNR filtering in filterReadsBam

ppapasaikas committed:
new tests for SNR calculation in scanGenome.

adjust tests for calcReadStats to account for additional metadata related to SNR calculation

ppapasaikas commented:
I looked at the generated reports but can’t figure out what is failing…
Any ideas / help would be much appreciated:-)

csoneson commented:
Here's what I get locally: 

```
── Failed tests ──────────────────────────────────────────────────────────────────────────────────
Error (test-scanGenome.R:516:5): NoiseFilterParam and its validator work
<rlib_error_3_0/rlib_error/error/condition>
Error in `"fun(..., .envir = .envir)"`: ! Could not evaluate cli `{}` expression: `x`.
Caused by error in `eval(expr, envir = envir)`:
! object 'x' not found
Backtrace:
     ▆
  1. ├─testthat::expect_error(...) at test-scanGenome.R:516:5
  2. │ └─testthat:::expect_condition_matching(...)
  3. │   └─testthat:::quasi_capture(...)
  4. │     ├─testthat (local) .capture(...)
  5. │     │ └─base::withCallingHandlers(...)
  6. │     └─rlang::eval_bare(quo_get_expr(.quo), quo_get_env(.quo))
  7. ├─footprintR:::.validate_NoiseFilterParam(...)
  8. │ └─cli::cli_abort(...) at footprintR/R/scanGenome.R:612:9
  9. │   └─cli:::vcapply(message, format_inline, .envir = .envir)
 10. │     └─base::vapply(X, FUN, FUN.VALUE = character(1), ..., USE.NAMES = USE.NAMES)
 11. │       └─cli (local) FUN(X[[i]], ...)
 12. │         ├─cli::cli_fmt(...)
 13. │         │ └─cli:::cli__rec(expr)
 14. │         └─cli (local) fun(..., .envir = .envir)
 15. │           ├─cli:::cli__message(...)
 16. │           │ └─"id" %in% names(args)
 17. │           └─cli:::glue_cmd(..., .envir = .envir, .call = sys.call(), .trim = FALSE)
 18. │             └─cli:::glue(...)
 19. ├─cli (local) `<fn>`("x")
 20. │ ├─.transformer(expr, .envir) %||% character()
 21. │ └─cli (local) .transformer(expr, .envir)
 22. │   ├─eval(expr, envir = envir) %??% ...
 23. │   │ └─chain_error(expr, err, srcref = utils::getSrcref(sys.call()))
 24. │   │   └─base::withCallingHandlers(...)
 25. │   └─base::eval(expr, envir = envir)
 26. │     └─base::eval(expr, envir = envir)
 27. └─base::.handleSimpleError(...)
 28.   └─h(simpleError(msg, call))
 29.     └─throw_error(err, parent = e)
```

mbstadler commented:
The problem was the use of `cli_abort` (no need to declare any variable in the call, and also no need to paste/collapse elements of a vector, this is all taken care of by `cli`). I have pushed a fix.

There are some nice online articles that exemplify `cli`'s capabilities, in particular https://cli.r-lib.org/articles/semantic-cli.html

mbstadler committed:
fix the use of cli_abort

ppapasaikas commented:
> The problem was the use of `cli_abort` (no need to declare any variable in the call, and also no need to paste/collapse elements of a vector, this is all taken care of by `cli`). I have pushed a fix.
> 
> There are some nice online articles that exemplify `cli`'s capabilities, in particular https://cli.r-lib.org/articles/semantic-cli.html

Thanks Michael!
Do you think we could merge with Devel?

mbstadler commented:
I think we still have a few points open from the list above (e.g. test coverage, version bump, update of NEWS.md), and you had also asked me to look into the `filterReadsBam` C++ function, which I have not had time to do yet.

ppapasaikas commented:
I did not mean to rush this. I’ll check again your points from the list and try to close most of them.

ppapasaikas commented:
Besides merging will be a bit more painful that it should since now there will be conflicts with the sampleEntropy branch

ppapasaikas committed:
Added test for plotNoisePars

ppapasaikas committed:
Additional tests for snrScoreWindows, estimateNoiseParsWindows

ppapasaikas committed:
Update NEWS.md
Small edits in snrScoreWindows doc

ppapasaikas commented:
@mbstadler  An update on the points in the list:

> * a brief description of the background model would be great, at least for us internally? 
-- Done

> * there are many hard-coded/hidden cutoffs in the code. Should maybe some of them be exposed as arguments, so that we are aware of them and can adjust if necessary (for example the constants in `min_diffs <- max(16L, floor(0.05 * length(x)))` or `d_cut <- max(10, quantile(depth, 0.05, na.rm = TRUE))` in `.estimate_snr_vec`, or `d_cut <- max(10, quantile(depthVals, 0.05, na.rm = TRUE))` and the six lower and upper quantiles used for pre-fit filtering in `estimateNoiseParsWindows`)?
-- Now exposed and validated

> * alternatively, could a robust `lm` be an alternative to the standard `lm` that would eliminate many or all of these hard-coded parameters?
-- Trimming extremes works well with standard lm. Have not looked in detail if robust lm can work without any filtering but my guess is at least some level of trimming extreme low/high meanMeth values would still be needed.

> * should we rename `nWindows` to `nWindowsPerChromosome` in `estimateNoiseParsWindows()`, to better reflect its meaning?
-- nWindows is now total number windows so naming is accurate

> * would it be an option to lift-out the plotting functionality in `estimateNoiseParsWindows()` to a separate function, consistent with other parameter estimation/plotting functions (e.g. `calcReadStats`, `plotReadStats`)?
-- Done, in the sense that plotting is now a separate function. There's still a plotting option in the function though.

> * should we make the plot in `estimateNoiseParsWindows()` using `ggplot2` instead of base graphics, to be consistent with all other plotting?
-- Done

> * we should add some unit tests (we currently have ~200 missed lines in scanGenome.R)
-- Currently at 98% with 9 lines not covered. Will try to cover those as well.

> * we should also add `minSNR` parameter and filtering to `filterReadsBam`, which is in every other way consistent with `filterReads`
-- Done but with some  doubts on the cpp code

> * we should bump the version and update `NEWS.md`
-- Updated NEWS.md. Will wait before bumping version as I am not sure which merge will happen first (this or the SampleEntropy branch). At any rate there will be conflicts between the two

mbstadler commented:
That looks very good - I'll try to look into `filterReadsBam` as soon as possible

mbstadler commented:
I would suggest we focus on merging this PR first and thus bump the version to be consistent with `NEWS.md`, if that is ok for you.

It seems we only need:

- [ ] my feedback on `filterReadsBam`
- [ ] the version bump (consistent with `NEWS.md`)
- [x] unit tests for uncovered lines (currently 9)

ppapasaikas committed:
Added tests for snrScoreWindows, estimateNoiseParsWindows

ppapasaikas committed:
Added test for missing assays on estimateNoiseParsWindows

ppapasaikas committed:
Rcpp export compute_snr_na_gap_aware to allow testing

ppapasaikas committed:
Added condition on nrow(sePos) to return NAs

ppapasaikas committed:
test compute_snr_na_gap_aware

ppapasaikas committed:
test edgecases in plotReadStats and .estimate_snr_probList in calcReadStats

ppapasaikas committed:
Temporary hack due to the fact that in the case of single stats the stat names do not make it to the constructed df colnames

ppapasaikas committed:
remove qcCOL=NULL

ppapasaikas committed:
replace as.data.frame coercion with explicit rbind of list elements to preserve colnames in case of single stat.

ppapasaikas committed:
- group and group_name are not required so simplify.
- Add check that not both qcCol and readInfoCol are NULL and add test for this case
- Test for SEntrModProb can now use a single stat

ppapasaikas committed:
read SNR changes:
- SNR.cpp split to estimateNoise.cpp (noise estimation using lag1 deltas) + estimateSNR.cpp(SNR estimation using raw noise esimtates + background noise model
- Corresponding changes in filter_modbam_cpp, defaults changes
- in calcReadStats naming .estimate_snr_probList -> .estimateSNRprobList
- in calcReadStats .estimateSNRprobList uses the cpp functions
- Naming .estimate_snr_probList -> .estimateSNRprobList
- Numeric tessthat tests to test new SNR calculation identical to original
- Minor change in labelDists roxygen that was producing a warning

ppapasaikas committed:
Refactoring  and other changes on windowed SNR functions:
- use cpp code (estimateNoise and estimateSNR) for Noise and SNR estimation
- Function naming to include SNR/Noise consistently
- estimateNoiseParsWindoes does not plot. Returns optionally data fro plotting
- Corresponding changes in tests
- Remove constructor for filtering thresholds. Now part of the estimateNoiseParsWindows

ppapasaikas committed:
Minor SNR related cpp changes
estimateNoise and estimateSNR can return total variance if Noise cannot be estimated

ppapasaikas committed:
Refactor calcReadStats to avoid <<- :
The outer lapply returns, for each sample, a list with 'stats' and 'floor_pars'.
Outside lapply, build 'out' from the 'stats' components and metadata(out)$snr_noise_coef from the 'floor_pars' components.

ppapasaikas commented:
Notice that SNR.cpp is still there for "historical" purposes, mainly to confirm that the estimateNoise.cpp + estimateSNR.cpp refactoring returns identical results (it does). 
It will be removed before merging to devel.

ppapasaikas committed:
SNR behavior:

- estimateSNR returns also raw Noise estimate
- Remove from filter_modbam_cpp guards already covered in estimateNoise or estimateSNR
- estimateSNR produces error msg when baseline needed but non-finite betas
- estimateSNR returns NAs when baseling requiredbut features non-finite
- In floor mode if noiseRaw is non finite fall back to baseline instead of returning NA
- Tests to cover use cases for estimateSNR

mbstadler committed:
remove .gitignore from git-tracked files

Co-authored-by: Charlotte Soneson <charlottesoneson@gmail.com>
Co-authored-by: Michael Stadler <stadler.michael@gmail.com>

mbstadler committed:
object name changes suggested by lintr

Co-authored-by: Charlotte Soneson <charlottesoneson@gmail.com>
Co-authored-by: Michael Stadler <stadler.michael@gmail.com>

mbstadler committed:
install temporary fix for ComplexUpset/ggplot2 v4 from PR

Co-authored-by: Charlotte Soneson <charlottesoneson@gmail.com>
Co-authored-by: Michael Stadler <stadler.michael@gmail.com>

csoneson committed:
Re-indent filter_modbam_cpp.cpp

Co-authored-by: Michael Stadler <stadler.michael@gmail.com>
Co-authored-by: Charlotte Soneson <charlottesoneson@gmail.com>

csoneson committed:
Import specific functions from ggplot2 and SummarizedExperiment

csoneson committed:
Limit number of cores in test

csoneson committed:
Some cleanup and simplifications

Co-authored-by: Michael Stadler <stadler.michael@gmail.com>
Co-authored-by: Charlotte Soneson <charlottesoneson@gmail.com>

csoneson committed:
Bump version

Co-authored-by: Michael Stadler <stadler.michael@gmail.com>
Co-authored-by: Charlotte Soneson <charlottesoneson@gmail.com>

csoneson committed:
Add new functions to _pkgdown.yml

Co-authored-by: Michael Stadler <stadler.michael@gmail.com>
Co-authored-by: Charlotte Soneson <charlottesoneson@gmail.com>
